src/System.Xml.XDocument/System/Xml/Linq/XContainer.cs

```diff
@@ -851,10 +853,74 @@ internal static string GetStringValue(object value)
         internal void ReadContentFrom(XmlReader r)
         {
             if (r.ReadState != ReadState.Interactive) throw new InvalidOperationException(SR.InvalidOperation_ExpectedInteractive);
-            XContainer c = this;
+
+            ContentReader cr = new ContentReader(this);
+            while (cr.ReadContentFrom(this, r) && r.Read()) ;
+        }
+
+        internal void ReadContentFrom(XmlReader r, LoadOptions o)
+        {
+            if ((o & (LoadOptions.SetBaseUri | LoadOptions.SetLineInfo)) == 0)
+            {
+                ReadContentFrom(r);
+                return;
+            }
+            if (r.ReadState != ReadState.Interactive) throw new InvalidOperationException(SR.InvalidOperation_ExpectedInteractive);
+
+            ContentReader cr = new ContentReader(this);
+            while (cr.ReadContentFrom(this, r, o) && r.Read()) ;
+        }
+
+        internal async Task ReadContentFromAsync(XmlReader r, CancellationToken token)
+        {
+            if (r.ReadState != ReadState.Interactive) throw new InvalidOperationException(SR.InvalidOperation_ExpectedInteractive);
+
+            ContentReader cr = new ContentReader(this);
+            while (!token.IsCancellationRequested && cr.ReadContentFrom(this, r) && await r.ReadAsync().ConfigureAwait(false)) ;
+
+            token.ThrowIfCancellationRequested();
+        }
+
+        internal async Task ReadContentFromAsync(XmlReader r, LoadOptions o, CancellationToken token)
+        {
+            if ((o & (LoadOptions.SetBaseUri | LoadOptions.SetLineInfo)) == 0)
+            {
+                await ReadContentFromAsync(r, token).ConfigureAwait(false);
+                return;
+            }
+            if (r.ReadState != ReadState.Interactive) throw new InvalidOperationException(SR.InvalidOperation_ExpectedInteractive);
+
+            ContentReader cr = new ContentReader(this);
+            while (!token.IsCancellationRequested && cr.ReadContentFrom(this, r, o) && await r.ReadAsync().ConfigureAwait(false)) ;
+
+            token.ThrowIfCancellationRequested();
```

Comment:
In general, cancellation should not be checked for / cause an exception after the relevant work has already completed.  This loop should be restructured such that ThrowIfCancellationRequested occurs where you currently poll IsCancellationRequested.